File: src/api/routes/learning_outcomes.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from src.api.deps import get_db
from src.schemas.learning_outcome import (
    LearningOutcomeCreate,
    LearningOutcomeResponse,
    LearningOutcomeUpdate,
)
from src.services import learning_outcome_service

router = APIRouter()


@router.get("/courses/{course_id}/learning-outcomes", response_model=dict)
def list_learning_outcomes(course_id: int, db: Session = Depends(get_db)):
    learning_outcomes = learning_outcome_service.list_learning_outcomes(db, course_id)
    return {
        "data": [LearningOutcomeResponse.model_validate(item) for item in learning_outcomes],
        "message": "Learning outcomes loaded",
    }


@router.post(
    "/courses/{course_id}/learning-outcomes",
    status_code=status.HTTP_201_CREATED,
    response_model=dict,
)
def create_learning_outcome(
    course_id: int,
    payload: LearningOutcomeCreate,
    db: Session = Depends(get_db),
):
    learning_outcome = learning_outcome_service.create_learning_outcome(db, course_id, payload)
    return {
        "data": LearningOutcomeResponse.model_validate(learning_outcome),
        "message": "Learning outcome created",
    }


@router.put("/learning-outcomes/{learning_outcome_id}", response_model=dict)
def update_learning_outcome(
    learning_outcome_id: int,
    payload: LearningOutcomeUpdate,
    db: Session = Depends(get_db),
):
    learning_outcome = learning_outcome_service.update_learning_outcome(db, learning_outcome_id, payload)
    return {
        "data": LearningOutcomeResponse.model_validate(learning_outcome),
        "message": "Learning outcome updated",
    }


@router.delete("/learning-outcomes/{learning_outcome_id}", response_model=dict)
def delete_learning_outcome(learning_outcome_id: int, db: Session = Depends(get_db)):
    learning_outcome_service.delete_learning_outcome(db, learning_outcome_id)
    return {
        "data": {"id": learning_outcome_id},
        "message": "Learning outcome deleted",
    }

File: frontend/src/api/learningOutcomes.ts
import { api } from './client';
import {
  mockCreateLearningOutcome,
  mockDeleteLearningOutcome,
  mockGetLearningOutcomes,
  mockUpdateLearningOutcome,
} from '@/mocks/learningOutcomes';
import type { LearningOutcome, LearningOutcomePayload } from '@/types/learningOutcome';

interface ApiResponse<T> {
  data: T;
  message: string;
}

const useMock = import.meta.env.VITE_USE_MOCK === 'true';

export async function getLearningOutcomes(courseId: number): Promise<LearningOutcome[]> {
  if (useMock) {
    return mockGetLearningOutcomes(courseId);
  }

  const response = await api.get<ApiResponse<LearningOutcome[]>>(`/api/courses/${courseId}/learning-outcomes`);
  return response.data.data;
}

export async function createLearningOutcome(
  courseId: number,
  payload: LearningOutcomePayload,
): Promise<LearningOutcome> {
  if (useMock) {
    return mockCreateLearningOutcome(courseId, payload);
  }

  const response = await api.post<ApiResponse<LearningOutcome>>(
    `/api/courses/${courseId}/learning-outcomes`,
    payload,
  );
  return response.data.data;
}

export async function updateLearningOutcome(
  id: number,
  payload: LearningOutcomePayload,
): Promise<LearningOutcome> {
  if (useMock) {
    return mockUpdateLearningOutcome(id, payload);
  }

  const response = await api.put<ApiResponse<LearningOutcome>>(`/api/learning-outcomes/${id}`, payload);
  return response.data.data;
}

export async function deleteLearningOutcome(id: number): Promise<void> {
  if (useMock) {
    return mockDeleteLearningOutcome(id);
  }

  await api.delete<ApiResponse<{ id: number }>>(`/api/learning-outcomes/${id}`);
}
